### PR TITLE
Make alert messages consistent across languages

### PR DESCRIPTION
**Pull request by erik-krogh**
The tests fail and I know it. The internal PR fixes the tests (see backref towards the bottom).  

I recently added [a QL-for-QL query for detecting inconsistent alert messages across languages](https://github.com/github/codeql/pull/10053). 

This PR fixes most of the alerts found by that query such that the queries now use mostly the same message across languages.   
I fixed the alerts by looking at the alert-message across all languages, and then I picked the message I liked the most.   
That choice is sometimes somewhat arbitrary, so feedback is very welcome. 
  
There are still some queries that use different messages, e.g. because they report a different level of detail, or because different alert messages better suit each language.    
(Example: [cpp really likes to use a callchain](https://github.com/github/codeql/blob/0ac8b7ce650bd4c0ccb085aa68286ba7ed5df9f4/cpp/ql/src/Security/CWE/CWE-022/TaintedPath.ql#L63), and the alert message is therefore different from the other languages).  

I fixed the alerts while developing the QL-for-QL query, and a lot of the fixed alert messages are not reported by the current query.  
I tried to get the QL-for-QL query to less noisy, but I could only manage that if I also removed some TPs, so I might make the QL-for-QL query more noisy in the future.

### CODE REVIEW

**Review by esbena — DISMISSED** [review]
JS 👍🏻

File: java/ql/src/DeadCode/UselessParameter.ql [hunk 1]
```diff
@@ -14,4 +14,4 @@ import semmle.code.java.deadcode.DeadCode
 
 from RootdefCallable c
 where not c.whitelisted()
-select c.unusedParameter() as p, "The parameter " + p + " is unused."
+select c.unusedParameter() as p, "The parameter '" + c + "' is never used."
```
Comment (aschackmull) [inline java/ql/src/DeadCode/UselessParameter.ql:17]:
```suggestion
select c.unusedParameter() as p, "The parameter '" + p + "' is never used."
```

File: java/ql/src/Violations of Best Practice/Dead Code/UnusedLocal.ql [hunk 2]
```diff
@@ -29,6 +29,4 @@ where
   // Rules about catch clauses belong in an exception handling query
   not exceptionVariable(ve) and
   not enhancedForVariable(ve)
-select v,
-  "Unused local variable " + v.getName() +
-    ". The variable is never read or written to and should be removed."
+select v, "The value assigned to local variable '" + v.getName() + "' is never used."
```
Comment (aschackmull) [inline java/ql/src/Violations of Best Practice/Dead Code/UnusedLocal.ql:32]: The new message says that the variable is being assigned a value, but the old message specifically says that it wan't written to. Which is it?

**Review by aschackmull — COMMENTED** [review]
Java looks mostly good to me - two minor comments, though.

**Review by RasmusWL — COMMENTED** [review]
Python :+1:

Overall question: Will changing these alert messages cause all current Code Scanning alerts to be closed, and new ones to pop up instead?

### Grammar nitpick
I think the new alert messages that read `User-provided value flows to here and is used in a command.` sounds odd grammatically. Shouldn't they be quantified to read `A user-provided value flows to here and is used in a command.`

Reading that sentence over a few times, I'm also starting to wonder about `flows to here`, which sounds odd, I would always say `flows here`. 

_(I not a gramtic genius, these just didn't match how I would talk with a person)_

Personally, I would suggest `A user-provided value flows here and is used in a command.`

**Review by smowton — DISMISSED** [review]
Go lgtm

File: java/ql/src/Violations of Best Practice/Dead Code/UnusedLocal.ql (hunk 2 above)
Comment (erik-krogh) [reply java/ql/src/Violations of Best Practice/Dead Code/UnusedLocal.ql:32]: It seems to be "wasn't written to". The query behaves slightly different compared to the other languages.  

I'll revert this change.

File: cpp/ql/src/Best Practices/BlockWithTooManyStatements.ql [hunk 3]
```diff
@@ -29,7 +29,4 @@ where
   n = strictcount(ComplexStmt s | s = b.getAStmt()) and
   n > 3 and
   complexStmt = b.getAStmt()
-select b,
-  "Block with too many statements (" + n.toString() +
-    " complex statements in the block). Complex statements at: $@", complexStmt,
-  complexStmt.toString()
+select b, "Block with too many statements (" + n.toString() + " complex statements in the block)."
```
[the remark this answers is not in the document]
Comment (erik-krogh) [reply cpp/ql/src/Best Practices/BlockWithTooManyStatements.ql:32]: Yes you're loosing which statements are being seen as complex.   
But the new message is the same as used by the other languages.  

And with the previous message you would have 1 alert pr. complex statement (so a minimum of 4).  
(Non LGTM alert showing is not that good at collapsing multiple alerts on the same line).   
So is that really better? (I'll revert this change if you still think so).

File: cpp/ql/src/Best Practices/Unused Entities/UnusedLocals.ql [hunk 4]
```diff
@@ -58,4 +58,4 @@ where
   not exists(AsmStmt s | f = s.getEnclosingFunction()) and
   not v.getAnAttribute().getName() = "unused" and
   not any(ErrorExpr e).getEnclosingFunction() = f // unextracted expr may use `v`
-select v, "Variable " + v.getName() + " is not used"
+select v, "The value assigned to local variable '" + v.getName() + "' is never used."
```
[the remark this answers is not in the document]
Comment (erik-krogh) [reply cpp/ql/src/Best Practices/Unused Entities/UnusedLocals.ql:61]: I used the alert message from Python to create this alert message for CPP.   
I'll do it the other way around instead 👍

**Comment by erik-krogh** [comment]
> Overall question: Will changing these alert messages cause all current Code Scanning alerts to be closed, and new ones to pop up instead?

Doesn't look like it. I tried to change an alert message [here](https://github.com/github/codeql/pull/10134), and code-scanning reported no change in alerts.  

> ### Grammar nitpick
> .... 
> Personally, I would suggest `A user-provided value flows here and is used in a command.`

There is a lot of those: https://cs.github.com/github/codeql?q=%22%5C%22User-provided+value%22  
Also beyond the queries I've touched here.   
So that seems like a nice idea for a followup PR.

File: cpp/ql/src/Best Practices/BlockWithTooManyStatements.ql (hunk 3 above)
[the remark this answers is not in the document]
Comment (geoffw0) [reply cpp/ql/src/Best Practices/BlockWithTooManyStatements.ql:32]: I'm also concerned by this change but also wondering if the original version was just a bit spammy.

It looks OK on LGTM: https://lgtm.com/query/382219361368642896/

File: java/ql/src/Likely Bugs/Comparison/BitwiseSignCheck.ql [hunk 5]
```diff
@@ -16,4 +16,4 @@ where
   e.isStrict() and
   e.getGreaterOperand() instanceof BitwiseExpr and
   e.getLesserOperand().(IntegerLiteral).getIntValue() = 0
-select e, "Sign check of a bitwise operation."
+select e, "Potential unsafe sign check of a bitwise operation."
```
Comment (aschackmull) [inline java/ql/src/Likely Bugs/Comparison/BitwiseSignCheck.ql:19]: Shouldn't that be
```suggestion
select e, "Potentially unsafe sign check of a bitwise operation."
```

Comment (smowton) [reply java/ql/src/Likely Bugs/Comparison/BitwiseSignCheck.ql:19]: Either works gramatically, with different bindings:

Potential (unsafe check)
(Potentially unsafe) check

Comment (aschackmull) [reply java/ql/src/Likely Bugs/Comparison/BitwiseSignCheck.ql:19]: Yeah, I just felt that the latter was closer to the intended message. But I guess both work semantically.

File: cpp/ql/src/Best Practices/BlockWithTooManyStatements.ql (hunk 3 above)
[the remark this answers is not in the document]
Comment (erik-krogh) [reply cpp/ql/src/Best Practices/BlockWithTooManyStatements.ql:32]: Yes. But LGTM is not staying around, so I think we should avoid the spammy message.

[the remark this answers is not in the document]
Comment (jketema) [reply cpp/ql/src/Best Practices/BlockWithTooManyStatements.ql:32]: I trust @geoffw0's judgement and experience here.

[the remark this answers is not in the document]
Comment (geoffw0) [reply cpp/ql/src/Best Practices/BlockWithTooManyStatements.ql:32]: > I trust @geoffw0's judgement and experience here.

If you look at the LGTM link above, though it collapses the duplicate results nicely, I don't think they're really _adding_ anything either.  The problem (complex statements) is plain to see in all of the cases.

Also, these kinds of stylistic queries have not been a priority for years.  If at some point we decide to pressure the Code Scanning team to improve presentation of duplicate results, its not going to be over this query!

[the remark this answers is not in the document]
Comment (jketema) [reply cpp/ql/src/Best Practices/BlockWithTooManyStatements.ql:32]: Yup, saw the LGTM output. Just wanted to make clear this no longer blocked by me anymore and my reasoning for that.

[the remark this answers is not in the document]
Comment (geoffw0) [reply cpp/ql/src/Best Practices/BlockWithTooManyStatements.ql:32]: Agreed.

**Review by aibaars — APPROVED** [review]

File: csharp/ql/src/Likely Bugs/UncheckedCastInEquals.ql [hunk 6]
```diff
@@ -35,4 +35,4 @@ where
   access = cast.getAChild() and
   access.getTarget().getDeclaringElement() = access.getEnclosingCallable() and
   nodeBeforeParameterAccess(access.getAControlFlowNode())
-select cast, "Missing type-check before casting parameter to 'Equals'."
+select cast, "Equals() method does not seem to check argument type."
```
Comment (tamasvajk) [inline csharp/ql/src/Likely Bugs/UncheckedCastInEquals.ql:38]: Shouldn't `@precision high` mean high confidence in the result? If so, then we could change `does not seem to check` to `does not check`.

**Review by tamasvajk — COMMENTED** [review]
Added a minor comment, otherwise C# LGTM.

**Review by nickrolfe — APPROVED** [review]
Changes since the previous approvals look good.

**Pull request merged**